amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -337,12 +341,26 @@ private static ShardFilter getShardFilterFromInitialPosition(InitialPositionInSt
     }
 
     private static List<Shard> getShardListAtInitialPosition(@NonNull final ShardDetector shardDetector,
-                                                     InitialPositionInStreamExtended initialPositionInStreamExtended) throws KinesisClientLibIOException {
+        InitialPositionInStreamExtended initialPositionInStreamExtended) throws KinesisClientLibIOException {
+
         final ShardFilter shardFilter = getShardFilterFromInitialPosition(initialPositionInStreamExtended);
-        final Optional<List<Shard>> shards = Optional.of(shardDetector.listShardsWithFilter(shardFilter));
+        List<Shard> shards;
+
+        for (int i = 0; i < retriesForCompleteHashRange; i++) {
+            shards = shardDetector.listShardsWithFilter(shardFilter);
 
-        return shards.orElseThrow(() -> new KinesisClientLibIOException("Stream is not in ACTIVE OR UPDATING state - " +
-                "will retry getting the shard list."));
+            if (shards == null) {
+                throw new KinesisClientLibIOException(
+                        "Stream is not in ACTIVE OR UPDATING state - will retry getting the shard list.");
+            }
+
+            if (hashRangeOfShardsIsComplete(shards)) {
+                return shards;
+            }
+        }
+
+        throw new KinesisClientLibIOException("Hash range of shards returned was incomplete after "
```

Comment:
Ended up deciding not to opt to use a list shards call, I'm not sure if it's a good idea for us to give the customer the option to shoot themselves in the foot if their stream has a large shard tree, I think that if we have an incomplete hash range from a position in the shard map then I'm not inclined to believe that a list shards call will return a complete shard range either; it seems like very unusual behavior for us to be returning a partial shard map repeatedly so I think it might be safest to just fast-fail, but we can sync on this if you think otherwise.